### PR TITLE
Integrate Optimized Slasher Into Beacon Node Runtime, Part 1

### PR DESCRIPTION
**Pull request by rauljordan**
Part of #8331, this PR integrates the new slasher code into the beacon node at runtime, able to run it using a --slasher flag. This involves starting a slasher process in the beacon node if the flag is enabled and sending over relevant feeds that slasher will receive data from during block and attestation validation pipelines.

This is the first of three PRs, and this one includes the following changes:

- [x] Forward all attestations in a block to the slasher in the blockchain service's process block routine
- [x] Forward attestations right after nil integrity checking in the sync, p2p validation pipeline to slasher
- [x] Forward block headers right after nil integrity checking in the sync, p2p validation pipeline
- [x] Register and start slasher as a service in the beacon node's node.go file
- [x] Initialize slasher by waiting for a chainstart notification to retrieve the genesis time
- [x] Define a `--slasher` feature flag used by the beacon node
- [x] Fix tests

In the next PR, we will modify our slashings pool to receive any slashings detected by the optimized slasher. In the final PR, we will verify the signatures of detected slashing events in slasher before sending any detected slashings over a feed.

### CODE REVIEW

File: beacon-chain/slasher/service.go [hunk 1]
```diff
@@ -58,15 +57,15 @@ func New(ctx context.Context, srvCfg *ServiceConfig) (*Service, error) {
...
 		ctx:                   ctx,
 		cancel:                cancel,
-		genesisTime:           time.Now(),
 	}, nil
 }
 
 // Start listening for received indexed attestations and blocks
 // and perform slashing detection on them.
 func (s *Service) Start() {
+	genesisTime := s.serviceCfg.GenesisTimeFetcher.GenesisTime()
```
Comment (rauljordan) [inline beacon-chain/slasher/service.go:66]: Slasher might need to wait until chain started?

File: beacon-chain/blockchain/service.go [hunk 2]
```diff
@@ -46,85 +47,91 @@ const headSyncMinEpochsAfterCheckpoint = 128
...
+	P2p                     p2p.Broadcaster
+	MaxRoutines             int
+	StateNotifier           statefeed.Notifier
+	ForkChoiceStore         f.ForkChoicer
+	OpsService              *attestations.Service
+	StateGen                *stategen.State
+	WspBlockRoot            []byte
+	WspEpoch                types.Epoch
+	SlasherBlockHeadersFeed *event.Feed
+	SlasherAttestationsFeed *event.Feed
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go:105]: Hard to see from the diff, but these are the only 2 new fields

File: beacon-chain/blockchain/service.go [hunk 3]
```diff
@@ -46,85 +47,91 @@ const headSyncMinEpochsAfterCheckpoint = 128
...
+		boundaryRoots:           [][32]byte{},
+		checkpointStateCache:    cache.NewCheckpointStateCache(),
+		opsService:              cfg.OpsService,
+		stateGen:                cfg.StateGen,
+		initSyncBlocks:          make(map[[32]byte]*ethpb.SignedBeaconBlock),
+		justifiedBalances:       make([]uint64, 0),
+		wsEpoch:                 cfg.WspEpoch,
+		wsRoot:                  cfg.WspBlockRoot,
+		slasherBlockHeadersFeed: cfg.SlasherBlockHeadersFeed,
+		slasherAttestationsFeed: cfg.SlasherAttestationsFeed,
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go:134]: These are the only 2 new fields

File: beacon-chain/node/node.go [hunk 4]
```diff
@@ -59,22 +60,26 @@ const testSkipPowFlag = "test-skip-pow"
...
+	exitPool                voluntaryexits.PoolManager
+	slashingsPool           slashings.PoolManager
+	depositCache            *depositcache.DepositCache
+	stateFeed               *event.Feed
+	blockFeed               *event.Feed
+	opFeed                  *event.Feed
+	slasherBlockHeadersFeed *event.Feed
+	slasherAttestationsFeed *event.Feed
+	proposerSlashingsFeed   *event.Feed
+	attesterSlashingsFeed   *event.Feed
```
Comment (rauljordan) [inline beacon-chain/node/node.go:80]: These are the only new fields

File: beacon-chain/sync/service.go [hunk 5]
```diff
@@ -50,17 +51,19 @@ var pendingBlockExpTime = time.Duration(params.BeaconConfig().SlotsPerEpoch.Mul(
...
+	AttPool                 attestations.Pool
+	ExitPool                voluntaryexits.PoolManager
+	SlashingPool            slashings.PoolManager
+	Chain                   blockchainService
+	InitialSync             Checker
+	StateNotifier           statefeed.Notifier
+	BlockNotifier           blockfeed.Notifier
+	AttestationNotifier     operation.Notifier
+	SlasherAttestationsFeed *event.Feed
+	SlasherBlockHeadersFeed *event.Feed
```
Comment (rauljordan) [inline beacon-chain/sync/service.go:65]: These are the only new fields

File: beacon-chain/sync/service.go [hunk 6]
```diff
@@ -120,24 +125,26 @@ func NewService(ctx context.Context, cfg *Config) *Service {
...
+		initialSync:             cfg.InitialSync,
+		attestationNotifier:     cfg.AttestationNotifier,
+		slotToPendingBlocks:     c,
+		seenPendingBlocks:       make(map[[32]byte]bool),
+		blkRootToPendingAtts:    make(map[[32]byte][]*ethpb.SignedAggregateAttestationAndProof),
+		stateNotifier:           cfg.StateNotifier,
+		blockNotifier:           cfg.BlockNotifier,
+		stateGen:                cfg.StateGen,
+		slasherAttestationsFeed: cfg.SlasherAttestationsFeed,
+		slasherBlockHeadersFeed: cfg.SlasherBlockHeadersFeed,
```
Comment (rauljordan) [inline beacon-chain/sync/service.go:146]: These are the only 2 new fields

File: beacon-chain/blockchain/process_block.go [hunk 7]
```diff
@@ -110,6 +112,39 @@ func (s *Service) onBlock(ctx context.Context, signed *ethpb.SignedBeaconBlock,
 		return err
 	}
 
+	// If slasher is configured, forward the attestations in the block via
+	// an event feed for processing.
+	if featureconfig.Get().EnableSlasher {
+		// Feed the indexed attestation to slasher if enabled. This action
+		// is done in the background to avoid adding more load to this critical code path.
+		go func() {
+			signedHeader, err := blockutil.SignedBeaconBlockHeaderFromBlock(signed)
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:121]: You are double importing block header here. We already do it in `validate_beacon_blocks.go`.

Are you thinking about initial syncing? if yes, a better place will be in `round_robin.go`

File: beacon-chain/slasher/service.go [hunk 8]
```diff
@@ -16,61 +18,91 @@ import (
...
 func (s *Service) Start() {
-	log.Info("Starting slasher")
-	s.slotTicker = slotutil.NewSlotTicker(s.genesisTime, params.BeaconConfig().SecondsPerSlot)
+	var genesisTime time.Time
+	stateChannel := make(chan *feed.Event, 1)
+	stateSub := s.serviceCfg.StateNotifier.StateFeed().Subscribe(stateChannel)
+	event := <-stateChannel
+
+	// Wait for us to receive the genesis time via a chain started notification.
+	if event.Type == statefeed.ChainStarted {
```
Comment (terencechain) [inline beacon-chain/slasher/service.go:67]: Are these conditions covered in unit tests? If not, we'll be getting poor test coverage

File: beacon-chain/sync/validate_beacon_attestation.go [hunk 9]
```diff
@@ -73,6 +75,32 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
 		return pubsub.ValidationReject
 	}
 
+	if featureconfig.Get().EnableSlasher {
+		// Feed the indexed attestation to slasher if enabled. This action
+		// is done in the background to avoid adding more load to this critical code path.
+		go func() {
+			preState, err := s.chain.AttestationPreState(ctx, att)
```
Comment (terencechain) [inline beacon-chain/sync/validate_beacon_attestation.go:82]: Note: This is a likely a dos vector. Someone could easily trick a slasher node to generate a state that uses more resources

File: beacon-chain/sync/validate_beacon_attestation.go [hunk 10]
```diff
@@ -73,6 +75,32 @@ func (s *Service) validateCommitteeIndexBeaconAttestation(ctx context.Context, p
...
+			if err != nil {
+				log.WithError(err).Error("Could not to retrieve pre state")
```
Comment (terencechain) [inline beacon-chain/sync/validate_beacon_attestation.go:84]:
```suggestion
				log.WithError(err).Error("Could not retrieve pre state")
```

File: beacon-chain/slasher/service.go (hunk 8 above)
Comment (rauljordan) [reply beacon-chain/slasher/service.go:67]: I can add some tests that check logs

**Review by terencechain — DISMISSED** [review]

**Review by terencechain — DISMISSED** [review]

**Pull request merged**